### PR TITLE
DHFPROD-6098: Create endpoint for fetching source document for mapping UI

### PR DESCRIPTION
**Pull request by brent-hartwig**
### Description
Offer the `/api/steps/mapping/{stepName}/newDoc` endpoint where the intent is to enable the mapping UI to retrieve a source document associated to a step.

Supports XML and JSON formatted source documents but only ever returns a JSON representation.  The implementation relies on fast-xml-parser to convert XML to JSON.  Version v3.17.1 thereof was introduced to the backend.  The frontend is already set up to use version ^3.16.0.

Whether in this PR or a subsequent one, this endpoint is *replace* `/api/steps/mapping/{stepName}/doc`.  When the PR was originally submitted, both endpoints were present.

#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [ ] JIRA_ID included in all the commit messages
- [ ] PR title is in the format JIRA_ID:Title
- [ ] Rebase the branch with upstream
- [ ] Squashed all commits into a single commit
- [ ] Added Tests
  

- ##### Reviewer:

- [ ] Reviewed Tests

### CODE REVIEW

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java [hunk 1]
```diff
@@ -83,6 +84,20 @@
         return new ResponseEntity<>(body, headers, HttpStatus.OK);
     }
 
+    @RequestMapping(value = "/{stepName}/newDoc", method = RequestMethod.GET)
+    @ApiOperation(value = "Returns an XML or JSON source document as a string of JSON")
+    @Secured("ROLE_readMapping")
+    public ResponseEntity<JsonNode> getNewDocument(@PathVariable String stepName, @PathVariable String docUri) {
+        // Step defines which database to pull the identified doc from.
+        final JsonNode stepNode = newService().getStep(STEP_DEFINITION_TYPE, stepName);
+        final String databaseName = stepNode.get("sourceDatabase").asText();
```
Comment (brent-hartwig) [inline marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java:93]: Is it okay to hard-code "sourceDatabase" (part of the step schema)?  I didn't find it used anywhere else within marklogic-data-hub-central.

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java [hunk 2]
```diff
@@ -83,6 +84,20 @@
...
+        // In the hub context, we're presuming staging or final.
+        final DatabaseClient databaseClient = databaseName.toLowerCase().equals(getHubClient().getDbName(DatabaseKind.STAGING).toLowerCase()) ?
+            getHubClient().getStagingClient() :
+            getHubClient().getFinalClient();
+        return ResponseEntity.ok(MappingService.on(databaseClient).getNewDocument(stepName, docUri));
```
Comment (brent-hartwig) [inline marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java:98]: The other endpoints herein use `StepService`.  Is it okay that this method uses `MappingService`?

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java [hunk 3]
```diff
@@ -83,6 +84,20 @@
         return new ResponseEntity<>(body, headers, HttpStatus.OK);
     }
 
+    @RequestMapping(value = "/{stepName}/newDoc", method = RequestMethod.GET)
+    @ApiOperation(value = "Returns an XML or JSON source document as a string of JSON")
+    @Secured("ROLE_readMapping")
+    public ResponseEntity<JsonNode> getNewDocument(@PathVariable String stepName, @PathVariable String docUri) {
+        // Step defines which database to pull the identified doc from.
+        final JsonNode stepNode = newService().getStep(STEP_DEFINITION_TYPE, stepName);
```
Comment (brent-hartwig) [inline marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java:92]: Should we determine the database from the step name here or on the backend, within `getNewDocument.sjs`?  The latter has the advantage of not presuming staging or final ...it's whatever the sourceDatabase property is.  From `getDocument.sjs`: `{database: xdmp.database(mappingStep.sourceDatabase)}`

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs [hunk 4]
```diff
@@ -0,0 +1,311 @@
+/**
```
Comment (brent-hartwig) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs:1]: Is this a good directory for this file?

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs [hunk 5]
```diff
@@ -0,0 +1,183 @@
+const test = require("/test/test-helper.xqy");
+const DataHubSingleton = require('/data-hub/5/datahub-singleton.sjs');
+const dataHub = DataHubSingleton.instance();
+const hubTest = require("/test/data-hub-test-helper.sjs");
+
+const jsonUri = '/test/sampleCustomerDoc.json';
```
Comment (brent-hartwig) [inline marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs:6]: Should all test data be moved out of this file?

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java (hunk 1 above)
Comment (brent-hartwig) [reply marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java:93]: No longer relevant given preceding answer from Ryan.

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java (hunk 3 above)
Comment (rjrudin) [reply marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java:92]: Yeah definitely have the backend figure it out.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.api [hunk 6]
```diff
@@ -0,0 +1,18 @@
+{
+    "functionName": "getNewDocument",
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.api:2]: This endpoint is returning a JSON object with 3 things - the actual document, an optional namespaces object, and a format string. I think we need a noun other than "document" here, as "document" implies you're just getting back a document, but you're getting more than that. 

The use case is - I want to test my mapping against a document. How about "getDocumentForTesting"? I think that gives us the semantic flexibility to return a JSON object where "document" is one property (as opposed to "data", which is too vague).

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.sjs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+var stepName, uri;
+
+const rtn = {
+  data: null,
+  namespaces: {},
+  format: null
+}
+
+const doc = cts.doc(uri);
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.sjs:28]: If the doc isn't found, we should immediately throw an error. Use ds-util.sjs, and add a new function to it that we haven't needed yet, but we need now - ds.throwNotFound("Could not find a document with URI: " + uri);

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+const textPropName = textStartsWith + 'text';
+const attrStartsWith = '@';
+const outputAttrStartsWith = '@'; // in case this ever needs to vary from the input.
+const nsPrefixDelim = ':';
+const nsAttrStartsWith = attrStartsWith + 'xmlns';
+const defaultNSAttrName = nsAttrStartsWith;
+const nonDefaultNSAttrStartsWith = nsAttrStartsWith + nsPrefixDelim
+
+// Set up for parser
+const parser = require('/fast-xml-parser/src/parser.js');
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs:29]: Is all of this code coming from our UI? There's a ton of logic in here and I'm wondering how much automated test coverage we have of it.

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs (hunk 5 above)
Comment (rjrudin) [reply marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs:6]: Here's what I recommend:

1. Use JUnit tests against the DS-generated interface for the scenarios of - URI exists, and URI doesn't exist. 
1. Use ml-unit-test for all of the actual logic, but invoke a function that does all the logic once the result of cts.doc has been obtained

That allows you to write ml-unit-tests that don't persist anything, which leads to much simpler tests.

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            'Sparrow'
+          ]
+        }
+      }
+    }
+  }
+};
+
+const xmlUri = '/test/sampleCustomerDoc.xml';
+const xmlInput = `
```
Comment (rjrudin) [inline marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs:28]: I think it'll be cleaner to have one test files for a JSON test, and one test file for an XML test. That's because the XML test is naturally going to be much more interesting. So e.g. getJsonDocumentForTestingTest.sjs and getXmlDocumentForTestingTest.sjs

**Review by rjrudin — CHANGES_REQUESTED** [review]
This looks good, but I definitely want to avoid adding all the fast-xml-parser files to source control. I'm assuming we need most of the files though? (I'm guessing we don't need the LICENSE file, e.g.)

I'll submit a PR into your PR that takes a different approach here. This is a new thing for us, and I want to get a good pattern setup for how we should reuse 3rd party libraries without storing all their source files in version control.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs (hunk 8 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs:29]: I wrote all of this code for this ticket.  I wasn't aware of existing code :/  Can you provide a pointer to that code?

Comment (srinathgit) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs:29]: The UI has some of the code (as the requirements for endpoint is slightly different) in mapping-card.tsx and source-to-entity-map.tsx

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+<Order xmlns="https://www.w3schools.com/OrderNS">
```
Comment (rjrudin) [inline marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs:29]: I recommend slimming this down to the bare minimum needed to verify that the XML parsing is working correctly. For example, you likely don't need all of the top-level properties that do not have a namespace - those are just adding noise.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs (hunk 8 above)
Comment (rjrudin) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs:29]: That's fine that you wrote it, I was just curious if it came from the UI first. 

The challenge here is there's a lot of code / conditionals / logic here, and I'm wondering how to ensure it's all covered correctly. There's one ml-unit-test for it, and while the XML input document has a lot of complexity to it, there's no way to know what the total set of scenarios is that we need to support. 

The tests can identify that though. Instead of having one large test that involves many scenarios, I think a better approach is to have a lot of small tests (they could all be separate functions in one file) that each test a single scenario. That would better identify what the total set of known scenarios that we support is.

Comment (brent-hartwig) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs:29]: Thanks, gents.  I'll check out the existing code (as it is proven) and will work over the tests as suggested.

**Comment by rjrudin** [comment]
Created https://github.com/brent-hartwig/marklogic-data-hub/pull/2 to avoid storing the fast-xml-parser files in source control. 

@brent-hartwig If that looks good, please modify the zip file to include the version number in the filename.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs (hunk 8 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs:29]: Initial assessment of referenced code...

### marklogic-data-hub-central/ui/src/components/entities/mapping/mapping-card.tsx:

* When source format is XML, fast-xml-parser's output is passed into generateNestedDataSource()
* When source format is JSON, that JSON is passed into generateNestedDataSource() as is.
* generateNestedDataSource() traverses the given JSON and creates a new structure that's quite a bit different than 6098's spec.
* Seems like only the traversal logic could be compared with DHFPROD-6098's implementation.

### /marklogic-data-hub-central/ui/src/components/entities/mapping/source-entity-map/source-to-entity-map.tsx

* Receives sourceData (output from generateNestedDataSource()) and fetchSrcDocFromUri (function) from mapping-card.tsx.
* fetchSrcDocFromUri appears only used when changing source document (via navigation).
* sourceData given to flattenSourceDoc()
* flattenSourceDoc() is the UI's implementation of DHFPROD-6101.

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java (hunk 2 above)
Comment (srinathgit) [reply marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java:98]: That's finr

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs (hunk 8 above)
Comment (srinathgit) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/xmlToJsonForMapping.sjs]: DHFPROD-6101 covers entirety of flattenSourceDoc() and some more. We will have to handle scenarios where properties are not Qnames in DHFPROD-6101.

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java (hunk 3 above)
Comment (brent-hartwig) [reply marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/controllers/steps/MappingStepController.java:92]: Addressed in fork; will be part of next or updated PR.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.api (hunk 6 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.api:2]: Changed to getDocumentForTesting.  Endpoint for front end also changed to `testingDoc`.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.sjs (hunk 7 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getNewDocument.sjs:28]: Addressed.  Will be present in updated/next PR.

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs (hunk 9 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs:28]: Done.  In the updated or next PR there will be a getDocumentForTesting sub-directory in the suites/data-hub/5/data-services/mapping directory.  At the time of this writing, it includes three files defining tests: nonExistentDocTest.sjs, startWithJson.sjs, and startWithXml.sjs.

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs (hunk 10 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs:29]: I was able to remove some elements that were not being tested, and added some tests for other unique portions that were not otherwise tested.

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs (hunk 5 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/mapping/getNewDocumentTest.sjs:6]: I added the JUnit tests as recommended, which surfaced a couple errors with that HC endpoint 👍.

**Comment by brent-hartwig** [comment]
This PR is ready for its second review.  Thank you in advance.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getDocumentForTesting.api [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{
+    "functionName": "getDocumentForTesting",
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getDocumentForTesting.api:2]: I know I proposed this, but seeing it here, I think it makes sense.

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/schemas/ExternalMetadata.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
+
```
Comment (rjrudin) [inline marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/schemas/ExternalMetadata.java:1]: Can you actually exclude this and any other schema-generated classes? I submitted a separate PR to update them.

File: marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/lib/mappingService.sjs [hunk 13]
```diff
@@ -0,0 +1,28 @@
+'use strict';
+
+// Utils for getDocumentForTesting's tests.
+const DocumentForTestingUtils = {
+  STEP_NAME: 'prospect2CustomerMappingStep',
+  DATABASE_NAME: 'data-hub-FINAL',
+  // Load data not provided via test-data directory.
+  loadTestData: function () {
+    // Found it difficult to figure out how to load a mapping step via test-data directory and get it into the collection
```
Comment (rjrudin) [inline marklogic-data-hub/src/test/ml-modules/root/test/suites/data-hub/5/data-services/lib/mappingService.sjs:9]: @srinathgit  Can you figure out the problem here? I know we have lots of ML tests that load mappings

**Review by rjrudin — CHANGES_REQUESTED** [review]
Looks good to me, just one comment about dropping ExternalMetadata, and hoping @srinathgit  can figure out the problem with loading mappings in a test. We want as little friction as possible with writing tests.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/ds-utils.sjs [hunk 14]
```diff
@@ -29,12 +29,17 @@ function throwForbidden(message) {
   fn.error(null, 'RESTAPI-SRVEXERR', Sequence.from([403, message]));
 }
 
+function throwNotFound(message) {
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/ds-utils.sjs:32]: Pinging @loriwilhelm  for awareness - this is the library where we're defining convenience functions for throwing errors from ML. As more of DHF is converted to use this library, we'll have an easy way of seeing where legit server error are being thrown.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getDocumentForTesting.api (hunk 11 above)
Comment (brent-hartwig) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getDocumentForTesting.api:2]: Does or does not make sense?

File: marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/schemas/ExternalMetadata.java (hunk 12 above)
Comment (brent-hartwig) [reply marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/schemas/ExternalMetadata.java]: Removed!

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getDocumentForTesting.api (hunk 11 above)
Comment (rjrudin) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getDocumentForTesting.api:2]: Does, I'm just patting myself on the back :)

**Review by rjrudin — APPROVED** [review]
Approving, and you can remove the "draft" status from it

**Review by srinathgit — CHANGES_REQUESTED** [review]
Can you squash all the commits to one and ensure all the tasks in checklist are done ? If they are all done, please go ahead and check them

**Comment by srinathgit** [comment]
@xnikhil08 
Can you review this PR and ensure that the response is in line with what we want for the UI ?

**Comment by brent-hartwig** [comment]
PR https://github.com/marklogic/marklogic-data-hub/pull/4835 is to replace this one.  It has all of this ticket's commits squashed into one.
